File: C++/Laba3sem2v2/hashTable.cpp
#include "hashTable.h"

#include <cstring>
#include <fstream>
#include <iostream>
#include <string>

using namespace std;

const int TABLE_SIZE = 10;
HashNode* hashTable[TABLE_SIZE] = {};

// -------------------------
// RAII-гвард дл€ очистки
// -------------------------
class HashTableGuard {
public:
    HashTableGuard() {
        initTable();   // очищаем таблицу при старте
    }
    ~HashTableGuard() {
        freeTable();   // очищаем таблицу при завершении программы
    }
};

// глобальный объект Ч работает автоматически
static HashTableGuard __hash_guard;


// -------------------------
// –еализаци€ функций
// -------------------------

void initTable() {
    freeTable(); // очищаем старые данные
    for (int i = 0; i < TABLE_SIZE; i++) {
        hashTable[i] = nullptr;
    }
}

int hashFunction(const string& key) {
    int hash = 0;
    for (char ch : key) hash += ch;
    return hash % TABLE_SIZE;
}

void insert(const string& key, const string& value) {
    int index = hashFunction(key);
    HashNode* current = hashTable[index];
    HashNode* prev = nullptr;

    while (current != nullptr) {
        if (current->key == key) {
            current->value = value;
            return;
        }
        prev = current;
        current = current->next;
    }

    HashNode* newNode = new HashNode{ key, value, nullptr };

    if (prev == nullptr)
        hashTable[index] = newNode;
    else
        prev->next = newNode;
}

string get(const string& key) {
    int index = hashFunction(key);
    HashNode* current = hashTable[index];

    while (current != nullptr) {
        if (current->key == key) return current->value;
        current = current->next;
    }
    return "NOT_FOUND";
}

void remove(const string& key) {
    int index = hashFunction(key);
    HashNode* current = hashTable[index];
    HashNode* prev = nullptr;

    while (current != nullptr && current->key != key) {
        prev = current;
        current = current->next;
    }

    if (current == nullptr) return;

    if (prev == nullptr)
        hashTable[index] = current->next;
    else
        prev->next = current->next;

    delete current;
}

void printTable() {
    for (int i = 0; i < TABLE_SIZE; i++) {
        cout << "Bucket " << i << ": ";
        HashNode* current = hashTable[i];
        while (current != nullptr) {
            cout << "[" << current->key << ": " << current->value << "] ";
            current = current->next;
        }
        cout << endl;
    }
}

void saveToFile(const string& fileName) {
    if (fileName.empty()) return;
    ofstream file(fileName);
    for (int i = 0; i < TABLE_SIZE; i++) {
        HashNode* current = hashTable[i];
        while (current != nullptr) {
            file << current->key << " " << current->value << endl;
            current = current->next;
        }
    }
}

void loadFromFile(const string& fileName) {
    if (fileName.empty()) return;
    freeTable();
    initTable();

    ifstream file(fileName);
    string key, value;
    while (file >> key >> value) insert(key, value);
}

void freeTable() {
    for (int i = 0; i < TABLE_SIZE; i++) {
        HashNode* current = hashTable[i];
        while (current != nullptr) {
            HashNode* temp = current;
            current = current->next;
            delete temp;
        }
        hashTable[i] = nullptr;
    }
}

void runHashTable(int argc, char* argv[]) {
    initTable();

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        }
        else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    if (!fileName.empty()) {
        loadFromFile(fileName);
    }

    string command;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    }
    else {
        command = query;
    }

    if (command == "HSET") {
        size_t pos = query.find(' ');
        string key = query.substr(0, pos);
        string value = query.substr(pos + 1);
        insert(key, value);
        if (!fileName.empty()) saveToFile(fileName);
    }
    else if (command == "HGET") {
        cout << get(query) << endl;
    }
    else if (command == "HDEL") {
        remove(query);
        if (!fileName.empty()) saveToFile(fileName);
    }
    else if (command == "HPRINT") {
        printTable();
    }
}

File: sem3l3go/hashtable/hashtable.go
package hashtable

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const tableSize = 10

// HashNode представляет узел хеш-таблицы
type HashNode struct {
	Key   string
	Value string
	Next  *HashNode
}

// HashTable представляет хеш-таблицу
type HashTable struct {
	table [tableSize]*HashNode
}

func (ht *HashTable) Clear() {
	for i := range ht.table {
		ht.table[i] = nil
	}
}

// NewHashTable создает новую хеш-таблицу
func NewHashTable() *HashTable {
	return &HashTable{
		table: [tableSize]*HashNode{},
	}
}

// hashFunction вычисляет хеш ключа
func (ht *HashTable) hashFunction(key string) int {
	hash := 0
	for _, ch := range key {
		hash += int(ch)
	}
	return hash % tableSize
}

// Insert вставляет или обновляет пару ключ-значение
func (ht *HashTable) Insert(key, value string) {
	index := ht.hashFunction(key)

	// Проверяем, существует ли уже ключ
	current := ht.table[index]
	for current != nil {
		if current.Key == key {
			current.Value = value
			return
		}
		current = current.Next
	}

	// Добавляем новый узел в начало
	ht.table[index] = &HashNode{
		Key:   key,
		Value: value,
		Next:  ht.table[index],
	}
}

// Get возвращает значение по ключу
func (ht *HashTable) Get(key string) string {
	index := ht.hashFunction(key)

	current := ht.table[index]
	for current != nil {
		if current.Key == key {
			return current.Value
		}
		current = current.Next
	}
	return "NOT_FOUND"
}

// Remove удаляет пару по ключу
func (ht *HashTable) Remove(key string) bool {
	index := ht.hashFunction(key)

	current := ht.table[index]
	var prev *HashNode

	for current != nil {
		if current.Key == key {
			if prev == nil {
				ht.table[index] = current.Next
			} else {
				prev.Next = current.Next
			}
			return true
		}
		prev = current
		current = current.Next
	}
	return false
}

// Print выводит содержимое таблицы
func (ht *HashTable) Print() {
	for i := 0; i < tableSize; i++ {
		fmt.Printf("Bucket %d: ", i)
		current := ht.table[i]
		for current != nil {
			fmt.Printf("[%s: %s] ", current.Key, current.Value)
			current = current.Next
		}
		fmt.Println()
	}
}

// SaveToFile сохраняет таблицу в файл
func (ht *HashTable) SaveToFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	for i := 0; i < tableSize; i++ {
		current := ht.table[i]
		for current != nil {
			if _, err := fmt.Fprintf(file, "%s:%s\n", current.Key, current.Value); err != nil {
				return err
			}
			current = current.Next
		}
	}
	return nil
}

// LoadFromFile загружает таблицу из файла
func (ht *HashTable) LoadFromFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())
		if len(parts) == 2 {
			ht.Insert(parts[0], parts[1])
		}
	}
	return scanner.Err()
}

// RunHashTable выполняет команды над хеш-таблицей
func RunHashTable(args []string) {
	ht := NewHashTable()
	var filename string
	var query string

	for i := 0; i < len(args); i++ {
		if args[i] == "--file" && i+1 < len(args) {
			filename = args[i+1]
			i++
		} else if args[i] == "--query" && i+1 < len(args) {
			query = args[i+1]
			i++
		}
	}

	if filename != "" {
		ht.LoadFromFile(filename)
	}

	parts := strings.SplitN(query, " ", 3)
	command := parts[0]
	var key, value string
	if len(parts) > 1 {
		key = parts[1]
	}
	if len(parts) > 2 {
		value = parts[2]
	}

	switch command {
	case "HSET":
		if key != "" && value != "" {
			ht.Insert(key, value)
			if filename != "" {
				ht.SaveToFile(filename)
			}
		}
	case "HGET":
		if key != "" {
			fmt.Println(ht.Get(key))
		}
	case "HDEL":
		if key != "" {
			ht.Remove(key)
			if filename != "" {
				ht.SaveToFile(filename)
			}
		}
	case "HPRINT":
		ht.Print()
	}
}
